Polyfills/WebSocket/Source/WebSocket.cpp

```diff
@@ -0,0 +1,146 @@
+            env,
+            JS_WEB_SOCKET_CONSTRUCTOR_NAME,
+            {
+                StaticValue("CONNECTING", Napi::Value::From(env, 0)),
+                StaticValue("OPEN", Napi::Value::From(env, 1)),
+                StaticValue("CLOSING", Napi::Value::From(env, 2)),
+                StaticValue("CLOSED", Napi::Value::From(env, 3)),
+                InstanceAccessor("readyState", &WebSocket::GetReadyState, nullptr),
+                InstanceAccessor("url", &WebSocket::GetURL, nullptr),
+                InstanceAccessor("onopen", nullptr, &WebSocket::SetOnOpen),
+                InstanceAccessor("onclose", nullptr, &WebSocket::SetOnClose),
+                InstanceAccessor("onmessage", nullptr, &WebSocket::SetOnMessage),
+                InstanceAccessor("onerror", nullptr, &WebSocket::SetOnError),
+                InstanceMethod("close", &WebSocket::Close),
+                InstanceMethod("send", &WebSocket::Send),
+            });
+
+        if (env.Global().Get(JS_WEB_SOCKET_CONSTRUCTOR_NAME).IsUndefined())
+        {
+            env.Global().Set(JS_WEB_SOCKET_CONSTRUCTOR_NAME, func);
+        }
+
+        JsRuntime::NativeObject::GetFromJavaScript(env).Set(JS_WEB_SOCKET_CONSTRUCTOR_NAME, func);
+    }
+
+    WebSocket::WebSocket(const Napi::CallbackInfo& info)
+        : Napi::ObjectWrap<WebSocket>{info}
+        , m_runtimeScheduler{JsRuntime::GetFromJavaScript(info.Env())},
+        m_webSocket{}
+    {
+        auto onOpenLambda = [this]()
+        {
+            m_runtimeScheduler([this]()
+            {
+                if(!m_onopen.IsEmpty())
+                {
+                    m_onopen.Call({});
+                }
+            });
+        };
+
+        auto onCloseLambda = [this]()
+        {
+            m_runtimeScheduler([this]()
+            {
+                Napi::Object closeEvent = Napi::Object::New(Env());
+                closeEvent.Set("code", 1000);
+                closeEvent.Set("reason", "CLOSED_EVENT NATIVE TEST CLOSE");
+                closeEvent.Set("wasClean", true);
+                if(!m_onclose.IsEmpty())
+                {
+                    m_onclose.Call({closeEvent});
+                }
+            });
+        };
+
+        auto onMessageLambda = [this](std::string message)
+        {
+            m_runtimeScheduler([this,message=std::move(message)]()
+            {
+                Napi::Object messageEvent = Napi::Object::New(Env());
+                messageEvent.Set("data", message);
+                if(!m_onmessage.IsEmpty())
+                {
+                    m_onmessage.Call({messageEvent});
+                }
+            });
+        };
+
+        auto onErrorLambda = [this]()
+        {
+            m_runtimeScheduler([this]()
+            {
+                Napi::Object errorEvent = Napi::Object::New(Env());
+                if(!m_onerror.IsEmpty())
+                {
+                    m_onerror.Call({errorEvent});
+                }
```

Comment:
Error object has no properties?

Reply:
Discussed offline, filed issue for this.